### ISSUE TITLE
List::Util min() doesn't work with objects overloading <

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7502]
Migrated from [rt.perl.org#31583](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=31583) (status was 'resolved')

Searchable as RT31583$

**Comment by p5pRT** [comment]
### From geoffr@gbdirect.co.uk

#### Created by geoffr@gbdirect.co.uk



In Perl 5\.6\.1 it was possible to pass objects \(such as DateTime values\)
into the min\(\) function from List&#8203;::Util \(and I would guess max\(\) is the
same\)\.  In Perl 5\.8 though it has been reimplemented in XS\, and this
just uses the numerical \< operator for comparison\, without checking
whether it actually needs to call a method if the operator has been
overloaded\.

The example below works fine with Perl 5\.6\.1\, but on Perl 5\.8\.4 it
produces warnings because it's trying to get a number out of the
stringified date to compare numerically\, and it appears to always
return the second value\.

&nbsp;    use warnings;
&nbsp;    use strict;
&nbsp;    use DateTime;
&nbsp;    use List&#8203;::Util qw\<min>;

&nbsp;    my $dt1 = DateTime\->new\(year => 2004\, month => 05\, day => 19\);
&nbsp;    my $dt2 = DateTime\->today;
&nbsp;    print '$dt1 \< $dt2 = '\, \($dt1 \< $dt2 ? 1 : 0\)\, "\\n";
&nbsp;    print 'min\($dt1\, $dt2\) = '\, min\($dt1\, $dt2\)\->ymd\, "\\n";

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl v5.8.4:

Configured by Debian Project at Wed May 12 00:32:16 EST 2004.

Summary of my perl5 (revision 5 version 8 subversion 4) configuration:
  Platform:
    osname=linux, osvers=2.4.26-ti1211, archname=i386-linux-thread-multi
    uname='linux kosh 2.4.26-ti1211 #1 sat apr 24 14:46:58 est 2004 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i386-linux -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.4 -Dsitearch=/usr/local/lib/perl/5.8.4 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.4 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.3.3 (Debian 20040429)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so.5.8.4
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.4:
    /etc/perl
    /usr/local/lib/perl/5.8.4
    /usr/local/share/perl/5.8.4
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.4:
    HOME=/home/geoffr
    LANG=en_GB
    LANGUAGE (unset)
    LC_ALL=en_GB
    LC_COLLATE=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/geoffr/usr/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/sbin:/usr/sbin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @gbarr

On 15 Sep 2004\, at 12&#8203;:59\, geoffr@&#8203;gbdirect\.co\.uk \(via RT\) wrote&#8203;:
> In Perl 5\.6\.1 it was possible to pass objects \(such as DateTime values\)
> into the min\(\) function from List&#8203;::Util \(and I would guess max\(\) is the
> same\)\.  In Perl 5\.8 though it has been reimplemented in XS\, and this
> just uses the numerical \< operator for comparison\, without checking
> whether it actually needs to call a method if the operator has been
> overloaded\.

List&#8203;::Util has always been implemented in XS\. However there is a perl
substitute that can be used if the compiled XS version is not installed\.

I suspect your 5\.6\.1 installation does not have a compiled XS version
installed and so you are seeing overload semantics in use\.

min\(\)/max\(\) in the XS were never designed to accept objects\.

Graham\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From nigelsandever@btconnect.com


> 
> In Perl 5\.6\.1 it was possible to pass objects \(such as DateTime values\)
> into the min\(\) function from List&#8203;::Util \(and I would guess max\(\) is the
> same\)\.  In Perl 5\.8 though it has been reimplemented in XS\, and this
> just uses the numerical \< operator for comparison\, without checking
> whether it actually needs to call a method if the operator has been
> overloaded\.
> 
> The example below works fine with Perl 5\.6\.1\, but on Perl 5\.8\.4 it
> produces warnings because it's trying to get a number out of the
> stringified date to compare numerically\, and it appears to always
> return the second value\.
> 
>     use warnings;
>     use strict;
>     use DateTime;
>     use List&#8203;::Util qw\<min>;
> 
>     my $dt1 = DateTime\->new\(year => 2004\, month => 05\, day => 19\);
>     my $dt2 = DateTime\->today;
>     print '$dt1 \< $dt2 = '\, \($dt1 \< $dt2 ? 1 : 0\)\, "\\n";
>     print 'min\($dt1\, $dt2\) = '\, min\($dt1\, $dt2\)\->ymd\, "\\n";
> 
> 
> 

Would minstr\(\) and maxstr\(\) not serve your purpose?

If not\, you can alway implement your own in terms of reduce\(\)\.

**Comment by p5pRT** [comment]
### From geoffr@gbdirect.co.uk

On Wed\, Sep 15\, 2004 at 01&#8203;:43&#8203;:29PM \-0000\, Graham Barr via RT wrote&#8203;:
> \.\.\.
> 
> List&#8203;::Util has always been implemented in XS\. However there is a perl
> substitute that can be used if the compiled XS version is not installed\.
>
> I suspect your 5\.6\.1 installation does not have a compiled XS version
> installed and so you are seeing overload semantics in use\.

It was the version in Debian stable\.  It may have been something
else that just meant that the warning wasn't produced\.

> min\(\)/max\(\) in the XS were never designed to accept objects\.

Yes\, the docs do say that it's "numerical" comparison\, but I first
took that to mean "the \< operator rather than lt"\.  It would be
nice if they did work with objects\, but if that's too hard to
implement then it might be a good idea to put a more explicit warning
in the docs\.

**Comment by p5pRT** [comment]
### From geoffr@gbdirect.co.uk

On Wed\, Sep 15\, 2004 at 03&#8203;:08&#8203;:17PM \-0000\, nigelsandever@&#8203;btconnect\.com via RT wrote&#8203;:
> 
> Would minstr\(\) and maxstr\(\) not serve your purpose?

Those happen to work in this particular case\, but I think it would
be better to make min\(\) and max\(\) a bit more polymorphic\.

> If not\, you can alway implement your own in terms of reduce\(\)\.

I did\, once I'd figured out what was going on\.

**Comment by p5pRT** [comment]
### From tels@bloodgate.com

\-\-\-\-\-BEGIN PGP SIGNED MESSAGE\-\-\-\-\-

Moin\,

Graham wrote&#8203;:
>On 15 Sep 2004\, at 12&#8203;:59\, geoffr@&#8203;gbdirect\.co\.uk \(via RT\) wrote&#8203;:
>> In Perl 5\.6\.1 it was possible to pass objects \(such as DateTime values\)
>> into the min\(\) function from List&#8203;::Util \(and I would guess max\(\) is the
>> same\)\.  In Perl 5\.8 though it has been reimplemented in XS\, and this
>> just uses the numerical \< operator for comparison\, without checking
>> whether it actually needs to call a method if the operator has been
>> overloaded\.
>
>List&#8203;::Util has always been implemented in XS\. However there is a perl
>substitute that can be used if the compiled XS version is not installed\.
>
>I suspect your 5\.6\.1 installation does not have a compiled XS version
>installed and so you are seeing overload semantics in use\.
>
>min\(\)/max\(\) in the XS were never designed to accept objects\.

I think the XS version should be fixed\, because it breaks in subtle ways
with bignum\. At first I thought it works by accident\, but it does so only
for small numbers&#8203;:

perl \-MList&#8203;::Util=min \-le 'print min\(2\, 3\)'
2
perl \-Mbignum \-MList&#8203;::Util=min \-le 'print min\(2\, 3\)'
2
perl \-Mbignum \-MList&#8203;::Util=min \-le 'print min\(2\*\*65\,\(2\*\*65\)\-1\)\," "\, 2\*\*65'
36893488147419103232 36893488147419103232

The XS vs\. Perl is a technical implementation detail and should not be 
visible to the user e\.g\. it should "just work" regardless on how it is 
implemented\.

Would it be simple \(what is ever simple?\) nec\. to add an overload/magic
check to the XS code?

Best wishes\,

Tels
\- \-\-
 Signed on Wed Sep 15 19&#8203;:52&#8203;:44 2004 with key 0x93B84C15\.
 Visit my photo gallery at http&#8203;://bloodgate\.com/photos/
 PGP key on http&#8203;://bloodgate\.com/tels\.asc or per email\.

 "Ivonna Humpalot"

\-\-\-\-\-BEGIN PGP SIGNATURE\-\-\-\-\-
Version&#8203;: GnuPG v1\.2\.4 \(GNU/Linux\)
Comment&#8203;: When cryptography is outlawed\, bayl bhgynjf jvyy unir cevinpl\.

iQEVAwUBQUiDA3cLPEOTuEwVAQHndQf7BCNKs0CShNRd6r4dNtwUxzSIuAVh/iR4
x1lVLWxeUGB\+\+gkbYRev0c9DOaYJGlZUSSg8AVnaUwtyfFDPC9F2LBJZennjNrxJ
K3EfyqkV1HBUh7P9Blhqb6lrDwCJp41qfOMALbAcTgr7/YXzaWxsB8U5KLd9Y4AD
kmFSBayqaxcLT/FpxffCmZ811MwRLVxuw2gR1AfXtPXx7Pu6J1tA8hfSya4B2Oom
6WcNaYDdaNcrcxdvZS3g9n/cqIL7E3BRVvco12Fmhu6dpngfZOuzsFDCFRUFbwgt
fQfz5us6mCjYQMbRPnaMmhPh0kbDkeb5NSeTauahMKMvDDcYI5My5Q==
=z0eV
\-\-\-\-\-END PGP SIGNATURE\-\-\-\-\-

**Comment by p5pRT** [comment]
### From @cpansprout

This was fixed in List&#8203;::Util 1\.20\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
